doc/sphinx/language/core/records.rst

```diff
@@ -1,248 +1,343 @@
       | {* @binder } := @term
-      term_record ::= %{%| {*; @field_def } {? ; } %|%}
-      field_def ::= @qualid {* @binder } := @term
-
-   Each :n:`@record_definition` defines a record named by :n:`@ident_decl`.
-   The constructor name is given by :n:`@ident`.
-   If the constructor name is not specified, then the default name :n:`Build_@ident` is used,
-   where :n:`@ident` is the record name. If given, the :n:`as @ident`
-   part specifies the name to use for inhabitants of the record in the
-   type of projections (see below).
-
-   If :token:`sort` is omitted, the default sort is Type.
-   Notice that the type of an identifier can depend on a previously-given identifier. Thus the
-   order of the fields is important. :n:`@binder` parameters may be applied to the record as a whole
-   or to individual fields.
-
-   .. todo
-      "Record foo2:Prop := { a }." gives the error "Cannot infer this placeholder of type "Type",
-      while "Record foo2:Prop := { a:Type }." gives the output "foo2 is defined.
-      a cannot be defined because it is informative and foo2 is not."
-      Your thoughts?
+      | {* @binder } @of_type := @term
+
+   Defines a non-recursive record type, creating projections for each field
+   that has a name other than `_`. The field body and type can depend on previous
+   fields, so the order of fields in the definition may matter.
+
+   Use the :cmd:`Inductive` and :cmd:`CoInductive` commands to define recursive
+   (inductive or coinductive) records.  These commands also permit defining
+   mutually recursive records provided that all of
+   the types in the block are records.  These commands automatically generate
+   induction schemes.  Enable the :flag:`Nonrecursive Elimination Schemes` flag
+   to enable automatic generation of elimination schemes for :cmd:`Record`.
+   See :ref:`proofschemes-induction-principles`.
 
    :n:`{? > }`
-     If provided, the constructor name is automatically declared as
+     If specified, the constructor is declared as
      a coercion from the class of the last field type to the record name
      (this may fail if the uniform inheritance condition is not
-     satisfied).  See :ref:`coercions`.
+     satisfied). See :ref:`coercions`.
+
+   :n:`@ident_decl`
+     The :n:`@ident` within is the record name.
+
+   :n:`{* @binder }`
+     :n:`@binder`\s may be used to declare the
+     :term:`inductive parameters <inductive parameter>` of the record.
+
+   :n:`: @sort`
+     The sort the record belongs to.  The default is :n:`Type`.
+
+   :n:`:= {? @ident }`
+     :n:`@ident` is the name of the record constructor.  If omitted,
+     the name defaults to :n:`Build_@ident` where :n:`@ident` is the record name.
+
+   :n:`as {? @ident}`
+     Specifies the name used to refer to the argument corresponding to the
+     record in the type of projections.  If not specified, the name is the first
+     letter of the record name.  See :ref:`example <record_as_clause>`.
```

Comment:
We could eventully add that is the lowercase form of the first letter of the type.

We could also add that in the case of a `Class`, it is by default the full name of the class. (This does seem a bad default choice actually, and, retrospectively, we could have done that for records also. Maybe too late to change and the `as` clause is now the way to have control on it.)

Reply:
I added a few words on `Class` with an example.  Perhaps it should be under the `Class` command description, but I think it's OK here for now.  (I think of `Record` as the primary command here and that the others are specializations (and thus secondary)).

One thing I don't understand is why there are so many ways to define records, i.e. `Record`, `Inductive`, `CoInductive`, `Class` and `Instance`.  I've not tried to understand typeclasses, so the need for the last two is a bit mysterious.  If one of you can provide a succinct way to distinguish those two, I'll add it in the text for the `Record` command where I mention `Inductive` and `CoInductive`.

Reply:
`Instance` inhabits a `Class`, so it is not a type but instead something equivalent to `{| .... |}`. `Class`es are `Record`s with extra information for automatic inference of inhabitants of the record. (Actually, it is a bit more subtle in that there are classes which are just definitions, that is, when a class is a record with only one field, we skip the indirection and treat the class as equal to the contents of the field; so technically `Class` is either a `Record` or a `Definition`. Again, when a Class is a Definition, it is to indicate extra information for automatic inference of the inhabitants).

When a `Record` is recursive, one need to know if it is `Inductive`, that is only with a finite nesting of the constructors (or more generally with only a well-founded nesting of the constructor), or `CoInductive`, that is with possibly infinite inhabitants (the prototypical example being streams: `CoInductive stream A := {hd:A; tl: stream A}.`).

When a record is not recursive, we don't need to tell if it is Inductive or CoInductive, so we use the word `Record` instead, to avoid having to arbitrarily choose an information Inductive/CoInductive which would be pointless.

Reply:
> I added a few words on Class with an example. Perhaps it should be under the Class command description, but I think it's OK here for now. (I think of Record as the primary command here and that the others are specializations (and thus secondary)).

I think that is ok as it is. This question of naming the inhabitant of the record or class is anyway relatively marginal compared to the logical aspects of a record.

Reply:
I'll just say this for now:

The :cmd:`Class` command can be used to define records with extra information that permits Coq to automatically infer the inhabitants of the record.